### ISSUE TITLE
Samtools view error reading and closing file "-" -5

### ISSUE THREAD

**Issue by kvn95ss** [samtools/samtools#1976]
#### Are you using the latest version of [samtools](https://github.com/samtools/samtools/releases/latest) and [HTSlib]
Samtools version is 1.13

 #### Please describe your environment.
- WSL and UPPMAX Rackham
#### Please specify the steps taken to generate the issue, the command you are running and the relevant output.

I am using this software to process our data - https://github.com/sdparekh/zUMIs

The pipeline is meant to be used with SmartSeq3 data (which is what I'm working with), however after alignment, the pipeline runs a command to essentially filter out reads based on some tag. 

The exact command which fails is - 

```
samtools view -h -@ 5 test.filtered.tagged.Aligned.out.bam  | grep -v -P 'UB:Z:\t' |  samtools view -b -@ 5 -o test.filtered.tagged.Aligned.out.bam.UMI.bam
samtools view -h -@ 5 test.filtered.tagged.Aligned.out.bam  | grep -v 'UB:Z:[A-Z]' |  samtools view -b -@ 5 -o test.filtered.tagged.Aligned.out.bam.internal.bam
```

The error I got in both case was 

```
samtools view: error reading file "-"
samtools view: error closing "-": -5
```

The Aligned.out.bam was generated from alignment using Star. The BAM file doesn't appear to be truncated - it passes ```samtools quickcheck``` with no issues, and here are the last 5 reads from the bam file

```
A00621:301:HTYHLDRXX:2:2269:19108:17613 0       chr6    42122567        0       2S14M62S        *       0       0       GCGGCAGGCCTAACACCTGTCTCTTATACACATCTCCGAGCCCACGAGACACTGAGCGATCTCGTATGCCGTCTTCTG  FFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFF:FFFFFFFFFFFFFFFFFFFFFFFF,F:F:FFFFF:FFFFF:F,FF NH:i:10 HI:i:1  AS:i:13 nM:i:0  BC:Z:EOIHXQFD   UB:Z:TACGGCTG   QB:Z:AAAAAAAA   QU:Z:FFFFFFFF
A00621:301:HTYHLDRXX:2:2269:23032:17707 16      chr11   133097171       3       54S24M  *       0       0       GGCATACGAGATCGCTCAGTGTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTATAAGAGACAGATTGCGCAATG  FFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF NH:i:2  HI:i:1  AS:i:15 nM:i:4  BC:Z:EOIHXQFD   UB:Z:AAGGGCCC   QB:Z:AAAAAAAA   QU:Z:FFFFFFFF
A00621:301:HTYHLDRXX:2:2269:13295:17754 16      chr9    136850871       255     29S71M  *       0       0       GTGGGCTCGGAGATGTGTATAAGAGACAGGCCTCCAGCAGCCACTTCAGAGCCCCCGCCTTTGCCTGCACTCCTCTTGCAGGGCTGGCCCTGCCTGCTCC   FFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF    NH:i:1  HI:i:1  AS:i:69 nM:i:0  BC:Z:EOIHXQFD   UB:Z:   QB:Z:AAAAAAAA   QU:Z:
A00621:301:HTYHLDRXX:2:2269:8314:17832  16      chr8    123558030       3       52S10M21945N16M *       0       0       AATACGAGATCGCTCAGTGTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGCTCCCATGGTGTGACGGGCGGTGTGT  ,FFFFF:F,FF,FFF:FFFFFFFFFF:F:FFFFFF:FFFFFFFF:FFFFFFF:FFF:,FF:FFFFFFFF,:F:FFFFF NH:i:2  HI:i:1  AS:i:18 nM:i:2  BC:Z:EOIHXQFD   UB:Z:TGCTCCGC   QB:Z:AAAAAAAA   QU:Z:,FFFF::F
A00621:301:HTYHLDRXX:2:2269:18602:17832 16      chr4    90880931        255     56S19M3S        *       0       0       ACGGCATACGAGATCGCTCAGTGTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTACTTAGATGTTTCAGTTCCC  :FFF:FFFFF:F:FFFFFFFFFFFFFFFFFFFFF:FFFF:FFFFFFFFFFFFFFFFFFFFFF,FFFFFFFFFF:FFFF NH:i:1  HI:i:1  AS:i:18 nM:i:0  BC:Z:EOIHXQFD   UB:Z:CAAACGGT   QB:Z:AAAAAAAA   QU:Z:FFFFFFFF
A00621:301:HTYHLDRXX:2:2269:32651:17863 16      chr9    136849417       255     16S84M  *       0       0       TTGCGCAATGAGGGACCCGGGAGGAGGAACATGGCGGTGGCGGACCTCGCTCTCATTCCTGATGTGGACATCGACTCCGACGGCGTCTTCAAGTATGTGC   ,FFFFFFF:FFFFFFFF:FFFFFFFFFFFF,:FFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFF    NH:i:1  HI:i:1  AS:i:82 nM:i:0  BC:Z:EOIHXQFD   UB:Z:   QB:Z:AAAAAAAA   QU:Z:
```

As such, the input and output file looks fine, what could the error be?

I tried running it locally, thinking it could be an I/O issue, but with my laptops SSD, 10 CPU and 20 GB ram, I still got the same error.

Any leads would be helpful!

**Comment by kvn95ss** [comment]
I rewrote the code to -

```
samtools view -h -@ 10 -e '[UB]==""' test.filtered.tagged.Aligned.out.bam  -o test.filtered.tagged.Aligned.out.bam.UMI.bam
samtools view -h -@ 10 -e '[UB]!=""' test.filtered.tagged.Aligned.out.bam  -o test.filtered.tagged.Aligned.out.bam.internal.bam
```
And now it works fine, the outputs appear fine using ```samtools quickcheck``` as well. Strange error to be honest :/

**Comment by daviesrob** [comment]
Using filter expressions is probably the best solution here, and likely much more efficient.  

I'm afraid with the information available it's not possible to say what went wrong with filtering the SAM format output, although it looks like the second `samtools view` in the pipeline was the one that failed.  It's possible that using a more recent version may give a better error report, assuming that it still can't read the output from `grep`.  And upgrading to the latest release is a good idea as lots of bugs have been fixed since version 1.13.

Note that `samtools quickcheck` can only find really bad problems with files, as it only looks at the beginning and end.  If there's a problem in the middle, it is unlikely to see anything wrong.  Also, `samtools` can write a reasonable-looking file, even if it fails for some reason.  It's best to check the exit code to ensure it worked properly.

**Comment by kvn95ss** [comment]
The problem continues...

The script then combines the bam file along with some more reads, and then sorts and indexes the bam. The sorting happens without any error, but when indexing I get the error - 

```
[E::hts_idx_push] Unsorted positions on sequence #171: 161676 followed by -9
[E::sam_index] Read 'A00621:301:HTYHLDRXX:2:2208:27724:2691' with ref_name='chrUn_GL000220v1', ref_length=161802, flags=0, pos=-9 cannot be indexed
```

This looks strange TBH, should I remove the offending read? I'm not sure what do :S
